cmd/cosign/cli/attach/rekor.go

```diff
@@ -0,0 +1,152 @@
+// limitations under the License.
+
+package attach
+
+import (
+	"context"
+	"crypto/sha256"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/sigstore/cosign/v2/cmd/cosign/cli/options"
+	"github.com/sigstore/cosign/v2/cmd/cosign/cli/rekor"
+	"github.com/sigstore/cosign/v2/cmd/cosign/cli/sign"
+	"github.com/sigstore/cosign/v2/pkg/cosign"
+	"github.com/sigstore/cosign/v2/pkg/cosign/bundle"
+	"github.com/sigstore/cosign/v2/pkg/oci/mutate"
+	ociremote "github.com/sigstore/cosign/v2/pkg/oci/remote"
+	"github.com/sigstore/cosign/v2/pkg/oci/static"
+	"github.com/sigstore/rekor/pkg/generated/client"
+	"github.com/sigstore/rekor/pkg/generated/models"
+)
+
+type tlogUploadFn func(*client.Rekor, []byte) (*models.LogEntryAnon, error)
+
+func uploadToTlog(ctx context.Context, sv *sign.SignerVerifier, rekorURL string, upload tlogUploadFn) (*bundle.RekorBundle, error) {
+	rekorBytes, err := sv.Bytes(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	rekorClient, err := rekor.NewClient(rekorURL)
+	if err != nil {
+		return nil, err
+	}
+	entry, err := upload(rekorClient, rekorBytes)
+	if err != nil {
+		return nil, err
+	}
+	fmt.Fprintln(os.Stderr, "tlog entry created with index:", *entry.LogIndex)
+	return bundle.EntryToBundle(entry), nil
+}
+
+func RekorCmd(ctx context.Context, regOpts options.RegistryOptions, rekorURL, sigRef, payloadRef, certRef, certChainRef, imageRef string) error {
+	b64SigBytes, err := signatureBytes(sigRef)
+	if err != nil {
+		return err
+	} else if len(b64SigBytes) == 0 {
+		return errors.New("empty signature")
+	}
+
+	ref, err := name.ParseReference(imageRef, regOpts.NameOptions()...)
+	if err != nil {
+		return err
+	}
+	ociremoteOpts, err := regOpts.ClientOpts(ctx)
+	if err != nil {
+		return err
+	}
+
+	digest, err := ociremote.ResolveDigest(ref, ociremoteOpts...)
+	if err != nil {
+		return err
+	}
+	// Overwrite "ref" with a digest to avoid a race where we use a tag
+	// multiple times, and it potentially points to different things at
+	// each access.
+	ref = digest // nolint
+
+	var payload []byte
+	if payloadRef == "" {
+		payload, err = cosign.ObsoletePayload(ctx, digest)
+	} else {
+		payload, err = os.ReadFile(filepath.Clean(payloadRef))
+	}
+	if err != nil {
+		return err
+	}
+
+	sig, err := static.NewSignature(payload, string(b64SigBytes))
+	if err != nil {
+		return err
+	}
+
+	var cert []byte
+	var certChain []byte
+
+	if certRef != "" {
+		cert, err = os.ReadFile(filepath.Clean(certRef))
+		if err != nil {
+			return err
+		}
+	}
+
+	if certChainRef != "" {
+		certChain, err = os.ReadFile(filepath.Clean(certChainRef))
+		if err != nil {
+			return err
+		}
+	}
+
+	sv, err := sign.SignerFromKeyOpts(ctx, certRef, certChainRef, options.KeyOpts{})
+	if err != nil {
+		return fmt.Errorf("getting signer: %w", err)
+	}
+	defer sv.Close()
+
+	bundle, err := uploadToTlog(ctx, sv, rekorURL, func(r *client.Rekor, b []byte) (*models.LogEntryAnon, error) {
```

Comment:
A bundle should already exist, correct? This command, like attach signature, should take an existing bundle and attach it to the container metadata.

Reply:
Thanks for the review, Yes even I felt the same initially but I was skeptical and had little less idea being new to the code base on how we will get rekor Bundle struct back? What should the user pass as a flag? Should rekor-url be passed (like it is passed with attest command?) or should we straight up use `sig.Bundle()`to see if it exists or not and if it does then write/mutate signature using withBundle() opts ?

Reply:
A rekor bundle is usually in the format: 
```json
{"Bundle":{"SignedEntryTimestamp":"MEUCIG...yoIY=","Payload":{"body":"...","integratedTime":1643917737,"logIndex":1,"logID":"4d2e4...97291"}
```

My only doubt is how will we allow the user to pass it to be attached? Or it hasn't has to be passed at all? And we can leverage `sig.Bundle()` because it already exists but was not attached in containers manifest (sounds contradicting to me as I am unsure about the passing thing)

Reply:
I would accept a path to a file with the bundle. `sig.Bundle` is where it's stored, and if it's set, it will be attached to the container.

Reply:
Got it so we will read that file and unmarshal it in the form of `*bundle.RekorBundle` and do something like this, right?:
```golang
	b, err := rekorBytes(rekorRef)
	if err != nil {
		return err
	}

	remoteOpts, err := regOpts.ClientOpts(ctx)
	if err != nil {
		return err
	}

	dstRef, err := ociremote.RekorTag(ref, remoteOpts...) //will have to create new func for RekorTag in ociremote 

	if err != nil {
		return err
	}

        //rough pseudo code only to tell the logic
	var bundle *bundle.RekorBundle
	json.Unmarshal(b, bundle)
	img, err := static.NewFile(b, static.WithBundle(bundle))
	if err != nil {
		return err
	}
```